Intro docs 79 (#29)
inrupt/docs-client-lib-js: source/tutorial/getting-started-part2.rst

The tutorial provides an explanation of the JavaScript code at the :ref:`end of the
tutorial <getting-started-code-details>`.

#. Create the :file:`my-demo.css` file with the following content:

   .. literalinclude:: /examples/getting-started/my-demo2.css
      :language: css
      :start-after: /* BEGIN CONTENT */

#. Create an :file:`index.html` file with the
   following content:

   .. literalinclude:: /examples/getting-started/index2.html
      :language: html
      :start-after: <!-- BEGIN CONTENT -->

#. Create an :file:`index.js` file with the
   following content:

   .. literalinclude:: /examples/getting-started/index2.js
      :language: javascript
      :start-after: // BEGIN CONTENT

   For details about the JavaScript code, see
   :ref:`getting-started-pt-2-code-details`.

5. Run the Application
----------------------

#. In the ``my-demo-app2`` directory, run:

   .. code-block:: sh

      npx parcel index.html

   The output resembles the following:

   .. code-block:: none

      Server running at http://localhost:1234
      ...
      Built in 2ms.

   If there are warnings with regards to the various source maps, you

>    can ignore them.

Reviewer: What errors are this? If it's something we can fix in the respective packages, that would be more professional than telling people to ignore warnings, IMHO.

Author: Heh heh -- I agree but I didn't know if you all would have time for that.   I'll retry and see later this morning.

Author: Removed the sentence.  I no longer get warnings (I think explicitly installing the libraries now instead of having it automatically do it when running did the trick).

#. Open ``localhost:1234`` in a browser.

#. Login.

   a. Click :guilabel:`Login` to login.

   #. If you have logged out of your Pod, you are prompted to log in.
      Enter your username and password and log in.

   #. If you have not already authorized the application, you will be
      prompted to authorize ``http://localhost:1234`` to access your
      Pod. To allow the application to read and write to your Pod,
      click :guilabel:`Authorize`.

   #. You are redirected back to your page.

   .. tip::

      - If you encounter redirect issues trying to log in, you may need

>         to clear the cache.

Reviewer: :O @NSeydoux Is this still true, i.e. not a bug that was fixed recently?

Author: Updated.  Will be in next patch

      .. literalinclude:: /examples/getting-started/index2.js
         :language: javascript
         :lines: 36

   .. tab:: Create Session

      Create a new :apisolidauthnbrowser:`Session
      </classes/_session_.session.html>`:

      .. literalinclude:: /examples/getting-started/index2.js
         :language: javascript
         :lines: 40

   .. tab:: Start Login

      Start the login process by calling :apisolidauthnbrowser:`Session.login()
      </classes/_session_.session.html#login>`:

      .. literalinclude:: /examples/getting-started/index2.js
         :language: javascript
         :lines: 47-55

      This function sends the user to the identity provider
      ``oidcIssuer``, and once logged in, redirects back to the specified
      ``redirectURL``.

   .. tab:: Finish Login

      When redirected back from the identity provider, use
      :apisolidauthnbrowser:`Session.handleIncomingRedirect()
      </classes/_session_.session.html#handleincomingredirect>` to
      complete the login process:

      .. literalinclude:: /examples/getting-started/index2.js
         :language: javascript
         :lines: 57-75

>       The function collects the information provided by the identity

Reviewer: ```suggestion
      The function verifies the information provided by the identity
```

@NSeydoux I think this might be more accurate?

.. tabs::

   .. tab:: Import

      From ``@inrupt/solid-client`` and ``@inrupt/vocab-common-rdf``,
      import the objects used in the application:

      .. literalinclude:: /examples/getting-started/index2.js
         :language: javascript
         :lines: 23-34,36,38

   .. tab:: Create SolidDataset

      Use :apisolidclient:`createSolidDataset
      </modules/_resource_soliddataset_.html#createsoliddataset>` to
      create a new SolidDataset (i.e., the reading list).

      .. literalinclude:: /examples/getting-started/index2.js
         :language: javascript
         :lines: 86

   .. tab:: Add Titles

      The application uses :apisolidclient:`createThing
      <modules/_thing_thing_.html#creatething>` to create a new data
      entity (i.e., :term:`Thing`) for each title to save. [#thingname]_

      To the Thing, the application uses the following functions to add
      data to the Thing:

      - :apisolidclient:`addUrl </modules/_thing_add_.html#addurl>` to
        add the ``RDF.type`` property with
        ``http://schema.org/article`` as its value.

      - :apisolidclient:`addStringNoLocale
        </modules/_thing_add_.html#addstringnolocale>` to add the

>         ``SCHEMA_INRUPT_EXT.name`` property with the title as its value.

Reviewer: Should we clarify why you need `SCHEMA_INRUPT_EXT`? (Because I have no idea...)

Reviewer: (Specifically, the `_INRUPT_EXT` part would strike me as curious if I were following this tutorial.)

Author: We probably should.  I debated myself.  On one hand, the lazy me was thinking "Well.... we kind of discuss the convenience objects in the getting started part 1 - but in a very light way.  will update.

Reviewer: I guess for the usage of constants in the first place that might be true, but then why can't we use regular `SCHEMA` - how does Inrupt extend it, and why does it matter? (Assuming that that's what `_INRUPT_EXT` means.)

Reviewer 2: `SCHEMA_INRUPT_EXT` is exactly that: an extension of the `schema.org` vocabulary, in particular adding language tags/translations to labels and comments lacking them in the canonical `schema.org`.

Reviewer: OK, so in that case... Regular plain old `SCHEMA` should suffice, right, since  we don't need tags/translations (or even labels and comments) here?

Reviewer 3: So this 'Inrupt extension' does more than just what Nic says - it also cherry-picks just a handful of terms from the 1,800 or so terms from Schema.org, thereby reducing the bundle size (and auto-complete noise, and just 'noise' in general). I thought there used to be a way to request subsets of terms (as Turtle) from Schema.org (e.g. to just request the Turtle for the Person class and its associated terms), but that doesn't seem to be available anymore.
So TL;DR: I'd just add a simple comment to that explain what SCHEMA_INRUPT_EXT is, and explain that it's an Inrupt extension to make working with select terms from Schema.org easier and quicker (i.e. rather than dealing with the clutter of all 1,800+ terms in there today).

Reviewer 3: And by the way, I removed the full-blown SCHEMA generated artifact from the generated bundle of common RDF vocabs. We could add it back easily at any time, or just offer it as it's own standalone bundle, but given the size of that bundle I didn't think anyone would really want to load all 1,800 vocab terms into their full-front apps.

Author: Updated.  Will be in next patch.

      Then, the application uses :apisolidclient:`setThing
      </modules/_thing_thing_.html#setthing>` to add the Thing to the
      SolidDataset (i.e., the reading list).

      .. literalinclude:: /examples/getting-started/index2.js
         :language: javascript
         :lines: 88-94

      The identifiers ``RDF.type``, and ``SCHEMA_INRUPT_EXT.name``, and

>       ``http://schema.org/article`` impose no restrictions on the data

Reviewer: What's the difference between `http://schema.org/article` and the others, i.e. why is it listed as a URL whereas the others are listed as JS object constants?

Author: Oh -  heh -- just an oversight.  I had used URLs for both and later thought I changed both, but clearly not.   Will update.

Author: Updated.

Author: Oh --so, it turns out there isn't SCHEMA_INRUPT_EXT.article (am looking at the installed SCHEMA_INRUPT_EXT.ts)  -- I only did a  grep and just looked whether it returned a match, not at what actually hit -- SCHEMA_INRUPT_EXT has a match for a  comment with the word article in it 😞   It should know what I meant.  ugh. Updated to use ActivityStream vocab.

      save the SolidDataset with the titles data to
      ``<PodURL>/getting-started/readingList/myListDS``.
      :apisolidclient:`saveSolidDatasetAt
      </modules/_resource_soliddataset_.html#savesoliddatasetat>`
      creates any intermediate folders/containers as needed. [#behavior]_

      .. literalinclude:: /examples/getting-started/index2.js
         :language: javascript
         :lines: 97-101
         :dedent: 4

      Upon successful save, :apisolidclient:`saveSolidDatasetAt
      </modules/_resource_soliddataset_.html#savesoliddatasetat>`
      returns a SolidDataset that reflects its state after the save.

      .. [#behavior]
      
         The ``solid-client`` library also provides the
         :apisolidclient:`saveSolidDatasetInContainer
         </modules/_resource_soliddataset_.html#savesoliddatasetincontainer>`. 
         However, unlike :apisolidclient:`saveSolidDatasetAt
         </modules/_resource_soliddataset_.html#savesoliddatasetat>`
         which creates any intermediate folders/containers as needed,
         :apisolidclient:`saveSolidDatasetInContainer
         </modules/_resource_soliddataset_.html#savesoliddatasetincontainer>`

>          requires the specified destination container already exists.

Reviewer: ```suggestion
         requires that the specified destination container already exists.
```

(I think this is needed, but not a native speaker...)

   .. tab:: Verify

      Upon save, the application re-retrieves the SolidDataset (the

>       reading list). The ``savedReadListDataSet`` returned from the
>       save operation reflects the saved state of the data. However, for

Reviewer: This is not exactly true: it represents the data that was sent to the Pod. This distinction would primarily be relevant when updating an existing Resource: because someone with write access (such as the current user in a different tab/device) might have also written data to the same Resource in between fetching and saving the data, the actual data in the Pod might be different from what was sent to it, and doing another fetch afterwards will ensure that you have the latest-available data.

It's hard to judge how often that will be something to take into account when writing an app, i.e. how detailed we should describe that here, but I suppose we at least shouldn't be saying it's the _saved_ data:

```suggestion
      reading list). The ``savedReadListDataSet`` returned from the
      save operation reflects the sent data. However, for
```

(This is something that we ran into recently, so this wasn't clear to myself either until recently.)

Author: Oh -- so, we currently have on the read/write data

```
Upon successful save, saveSolidDatasetAt returns a SolidDatset that reflects its state after the save. That means, if another process made separate non-conflicting modifications to the SolidDataset after you retrieved the SolidDataset but before you successfully saved, the returned SolidDataset reflects both the changes
```
I thought I tested this and it did return both changes, but could be remembering incorrectly.

Reviewer: Huh... Unless I'm misreading the source code, I don't think that is the case. I'll do a test and report back.

Author: Sounds good.  Will retest as well.

Reviewer: Just verified it, and indeed, the return value is not up-to-date with the saved Pod. [Test code.](https://codesandbox.io/s/optimistic-bose-f0ugm?file=/src/index.ts)

Screen recording of that same app running twice, with each instance editing different values in the same Resource:

![concurrent-updates](https://user-images.githubusercontent.com/4251/97895169-b9a3f600-1d33-11eb-934d-908feb32412b.gif)

Author: You are absolutely correct.  I can't recreate what I thought I saw last time (where I was changing the Role at inrupt.net while an application was updating the name).  Will update this and the other blurb.

Reviewer 3: Being totally predantic (as always!), but `savedReadListDataSet` shouldn't be capitalising `Set`, so it should be `savedReadListDataset`. Also I assume a `ReadList` is a list of things that I've already Read, as opposed to a `ReadingList`?

Author: Pat, Pat, Pat ... Haven't you been in a book club before?  One doesn't actually need to read things; it's enough just saving a list 😉   Heh -- updated.  And am removing  the Dataset suffix to these. Can put them back (with the appropriate spelling, of course).   Am of two minds - on one hand, it's nice to re-enforce the idea of a SolidDataset and then Things, etc.  But, then, the variable names seem so unwieldy.  Ugh.

      illustrative purposes, the tutorial uses
      :apisolidclient:`getSolidDataset
      </modules/_resource_soliddataset_.html#getsoliddataset>` to
      retrieve the SolidDataset again.

      .. literalinclude:: /examples/getting-started/index2.js
         :language: javascript
         :lines: 106-119
         :dedent: 4

      The application uses ``SCHEMA_INRUPT_EXT.name`` convenience object
      from the ``vocab-common-rdf`` library to specify the property to
      retrieve.

- For more information on property identifiers and vocabularies, see
  :doc:`/tutorial/use-vocabularies`.

- For more information on write operations, see :ref:`write-data`.

Additional Information
======================

API Documentation
-----------------

- `solid-client API
  <https://docs.inrupt.com/developer-tools/api/javascript/solid-client/README.html>`__

> - `solid-client-authn-browser API

Reviewer: @nicolasmondada should we be using the other names here as well?

         :language: javascript
         :lines: 89-95

>       ``RDF``, ``SCHEMA_INRUPT_EXT``, and ``AS`` provide convenience

Reviewer 3: There are repeated references to the vocabs `providing convenience objects`. Maybe you've discussed this before with the guys, but I'd suggest referring to them as `providing convenience constants`, and then this sentence might become `providing convenience constants that represent many of the ideas and concepts that are useful in Solid itself, and in Solid applications`.

Author: Heh - I believe convenience objects was actually your suggestion during the using vocabularies review :D 

> provide convenient objects (as JavaScript classes) containing static constants for common identifiers used right across Solid (and RDF in general). Importing these classes simply means developers don't need to hard-code these identifiers themselves in their code...

Heh - I had shortened it.  I'll footnote the details to break up the text.

      objects for many Solid-related terms:
      
      - ``RDF`` provides convenience objects for the `RDF Vocabulary
        <http://www.w3.org/1999/02/22-rdf-syntax-ns>`_. For example, the
        ``RDF.type`` is a convenience object that includes the
        ``http://www.w3.org/1999/02/22-rdf-syntax-ns#type`` IRI.

      - ``SCHEMA_INRUPT_EXT`` is Inrupt's extension of the `schema.org
        Vocabulary <http://schema.org>`_. It provides convenience
        objects for a **subset** of terms from the `schema.org
        Vocabulary`_, adding language
        tags/translations to labels and comments if missing from

>         schema.org.

Reviewer 3: I'd add a simple justification for **_why_** it's a subset - i.e. just append a sentence like "Schema.org itself currently defines over 2,500 terms (see [Schema.org](https://schema.org/docs/schemas.html)), whereas most applications (including Solid itself) really only require tiny specialized subsets of those terms. By using SCHEMA_INRUPT_EXT, which cherry-picks a small set of very generally applicable terms, we reduce noise, clutter and bundle sizes - but if you require a Schema.org term not currently cherry-picked by Inrupt in our extension, you can always declare it manually in your own code, create your own extension vocabulary, or request that Inrupt add that term to SCHEMA_INRUPT_EXT."

Author: Done - put  in the Vocabularies page with a shortened version in the getting started (with a reference to the SCHEMA_INRUPT_EXT) section.

What will probably happen (post-GA) is I'd like to create a reference page with these, so that when talking about FOAF, SCHEMA_INRUPT_EXT, and others as part of an example, we can just add 1-3 sentences in the context and have the readers go to a page that provides a fuller reference on these things.
